inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -40,190 +1230,466 @@ bool MKLDNNTopKNode::isSupportedOperation(const std::shared_ptr<ngraph::Node>& o
-void MKLDNNTopKNode::top1_axis(const float* src_data, float* dst_data, int* dst_idx, SizeVector in_dims) {
+void MKLDNNTopKNode::top1_axis(const float* src_data, float* dst_data, int32_t* dst_idx, SizeVector in_dims, std::function<float(float, float)> compare) {
     int after_num = count(in_dims, axis + 1, in_dims.size());
-    int first_index = 0;
-
-#if defined(HAVE_SSE) || defined(HAVE_AVX2) || defined(HAVE_AVX512F)
-    parallel_for2d(before_num, after_num / block_size, [&](int i0, int ib1) {
-            int s_index = i0 * dim * after_num + ib1 * block_size;
-            vec_type_f vmax_val = _mm_uni_loadu_ps(src_data + s_index);
-            vec_type_i vindex_max_val = _mm_uni_setzero_si();
-            for (int i2 = 1; i2 < dim; i2++) {
-                s_index += after_num;
-                vec_type_f vsrc = _mm_uni_loadu_ps(src_data + s_index);
-                vmask_type vmask = Compare1::cmp_ps(vsrc, vmax_val);
-                vmax_val = _mm_uni_blendv_ps(vmax_val, vsrc, vmask);
-
-                vec_type_i vindex_cur_val = _mm_uni_set1_epi32(i2);
-#if defined(HAVE_AVX512F)
-                vindex_max_val = _mm512_mask_blend_epi32(vmask, vindex_max_val, vindex_cur_val);
-#else
-                vindex_max_val = _mm_uni_blendv_epi8(vindex_max_val, vindex_cur_val, _mm_uni_castps_si(vmask));
-#endif
-            }
-            if (dst_data)
-                _mm_uni_storeu_ps(dst_data + i0 * after_num + ib1 * block_size, vmax_val);
-            if (dst_idx)
-                _mm_uni_storeu_si(reinterpret_cast<vec_type_i*>(dst_idx + i0 * after_num + ib1 * block_size), vindex_max_val);
-        });
-        first_index = after_num / block_size * block_size;
-#endif
-    int rest = after_num - first_index;
-    parallel_for2d(before_num, rest, [&](int i0, int i1) {
+
+    parallel_for2d(before_num, after_num, [&](int i0, int i1) {
         int index_max_val = 0;
-        int s_index = i0 * dim * after_num + first_index + i1;
+        int s_index = i0 * dim * after_num + i1;
         float max_val = src_data[s_index];
         for (int i2 = 1; i2 < dim; i2++) {
             s_index += after_num;
-            if (Compare2<float>()(src_data[s_index], max_val)) {
+            if (compare(src_data[s_index], max_val)) {
                 max_val = src_data[s_index];
                 index_max_val = i2;
             }
         }
         if (dst_data)
-            dst_data[i0 * after_num + first_index + i1] = max_val;
+            dst_data[i0 * after_num + i1] = max_val;
         if (dst_idx)
-            dst_idx[i0 * after_num + first_index + i1] = index_max_val;
+            dst_idx[i0 * after_num + i1] = index_max_val;
     });
 }
 
-template <template <typename> class Compare>
-void MKLDNNTopKNode::top1(const float* src_data, float* dst_data, int* dst_idx, SizeVector in_dims) {
+void MKLDNNTopKNode::top1(const float* src_data, float* dst_data, int32_t* dst_idx, SizeVector in_dims, std::function<float(float, float)> compare) {
```

Comment:
SizeVector can be replaced with const SizeVector&

Reply:
Method can be marked as const

Reply:
Fixed. Thanks Egor!